[mlir][linalg] Add a new helper hook: `hasVectorizationImpl`

banach-space opened the pull request:
The newly added hook simply returns `false` for Ops for which there's no
"vectorization logic" in the Linalg Vectorizer (i.e. the `vectorize()`
method). It's added so that the following two TD ops expose identical
level of functionality (that's not the case ATM):

  * `transform.structured.vectorize_children_and_apply_patterns`
  * `transform.structured.vectorize`

Specifically, ATM, the former works only for Linalg Ops, while the
latter works for all Ops that the vectorizer supports (*). With this change,
I am making sure that both TD will behave consistently.

Note, this shouldn't affect any of the current uses of the vectorizer.

(*) This is implemented via the `vectorize()` method in Vectorization.cpp.

banach-space committed:
[mlir][linalg] Add a new helper hook - `isVectorizable`

The newly added hook simply returns `false` for Ops for which there's no
"vectorization logic" in the Linalg Vectorizer (i.e. the `vectorize`
method). It's added so that the following two TD ops expose identical
level of functionality (that's not the case ATM):

  * `transform.structured.vectorize_children_and_apply_patterns`
  * `transform.structured.vectorize`

rengolin reviewed (commented):
I was expecting this would actually check if the op has the right method, perhaps driven by implementing an interface or something. The code here is mostly moving things round, NFC almost.

banach-space commented:
Yeah, I realise that the method name suggests something grand, but in reality is just a wrapper for an `if` stmt that I want to make sure is shared and re-used.

I hope that this is OK. Not strictly needed today, but I' working on a small refactor that this is blocking (as one of the TD ops bails out too early).

rengolin commented:
I don't mind the little snippet. @ftynse ?

hanhanW commented:
It is confusing because we have `vectorizeOpPrecondition` method; the real checks happen there.

https://github.com/llvm/llvm-project/blob/ce72c76e9bd0bf74af614cae6a9f85cfd4720e95/mlir/lib/Dialect/Linalg/Transforms/Vectorization.cpp#L2091-L2115

hanhanW commented on mlir/include/mlir/Dialect/Linalg/Transforms/Transforms.h:
```diff
@@ -762,6 +762,13 @@ LogicalResult copyToGPUPrivateMemory(OpBuilder &b, Value src, Value dst);
 /// memory is freed when going outside of the scope.
 LogicalResult deallocateGPUPrivateMemory(OpBuilder &, Value /*buffer*/);
 
+/// Check if this Op is vectorizable. All Linalg Op are vectorizable, as well
```
I think this statement is not correct. Not all the Linalg op are vectorizable. Linalg ops with non-trivial indexing maps are not vectorizable atm.

https://github.com/llvm/llvm-project/blob/ce72c76e9bd0bf74af614cae6a9f85cfd4720e95/mlir/lib/Dialect/Linalg/Transforms/Vectorization.cpp#L1916-L1926

banach-space commented:
> I don't mind the little snippet. @ftynse ?

Sure, but then we have to repeat the same snippet twice. Also, `isVectorizable` is really a property of the vectorizer that should be captured/documented somewhere. That's missing today.

In case it wasn't clear from the diff, this is what's happening today (pseudo-code to capture the high-level idea):

```cpp
// CURRENT logic for transform.structured.vectorize
LogicalResult process_vectorize_td(Operation *target) {
  if (!isa<linalg::LinalgOp, tensor::PadOp, tensor::PackOp, tensor::UnPackOp>(target))
    return rewriter.notifyMatchFailure(op, "Unsupported Op, cannot vectorize");

  return vectorize(target);
}

// CURRENT logic for transform.structured.vectorize_children_and_apply_patterns
LogicalResult process_vectorize_children_and_apply_patterns_td(Operation *target) {
  // NOTE - this condition is unnecessarily restrictive and is being relaxed in this PR
  if (!isa<linalg::LinalgOp>(target))
    return rewriter.notifyMatchFailure(op, "expected Linalg Op");

  return vectorize(target);
}
```

After this change, we will have this:
```cpp
bool isVectorizable(Operation *op) {
  return isa<linalg::LinalgOp, tensor::PadOp, tensor::PackOp, tensor::UnPackOp>(
      op);
}

// NEW logic for transform.structured.vectorize
LogicalResult process_vectorize_td(Operation *target) {
  if (!isVectorizable(target))
    return mlir::emitSilenceableFailure(target->getLoc())
           << "Unsupported Op, cannot vectorize";

  return vectorize(target);
}

// NEW logic for transform.structured.vectorize_children_and_apply_patterns
LogicalResult
process_vectorize_children_and_apply_patterns_td(Operation *target) {
  if (!isVectorizable(target))
    return mlir::emitSilenceableFailure(target->getLoc())
           << "Unsupported Op, cannot vectorize";

  return vectorize(target);
}
```

Put differently, this PR aims to achieve two things:
* healthier code re-use,
* relax the artificially imposed restrictions on `ransform.structured.vectorize_children_and_apply_patterns`.

banach-space commented:
> It is confusing because we have `vectorizeOpPrecondition` method; the real checks happen there.
> 
> https://github.com/llvm/llvm-project/blob/ce72c76e9bd0bf74af614cae6a9f85cfd4720e95/mlir/lib/Dialect/Linalg/Transforms/Vectorization.cpp#L2091-L2115

Yes, there are effectively two levels of verification (note that this PR doesn't change that). Here's pseudo-code to explain this:

```cpp
// Verification 1: At the level of transform.structured.vectorize (very crude, initial filtering)
LogicalResult process_vectorize_td(Operation *target) {
  if (!isVectorizable)
    return rewriter.notifyMatchFailure(op, "Unsupported Op, cannot vectorize");

  return vectorize(target);
}

// Verification 2: At the level of linalg::vectorize (very detailed)
LogicalResult vectorize(Operation *op) {
    if (failed(vectorizeOpPrecondition(op) {
      LDBG("Vectorization pre-conditions failed\n");
      return failure();
    }
    // ...
}
```

Basically:
* first, before doing anything expensive (and entering `vectorize()`), we check that the candidate Op is supported (*),
* second, for every Op that's supported, we run more involved checks inside `vectorize()`.

I can re-wire things, but would rather start with this small change :) I have a few other changes in flight that I'd like to prioritise (related to vectorising `tensor.pad`/`tensor.pack`).

(*) You are right that even some Linalg Ops don't qualify for vectorization (at least today). The idea behind `isVectorizable` (I should probably rename it) is to reject Ops that are known not to be supported at all, i.e. Ops outside the Linalg Dialect and selected Ops from Tensor. And for the supported Ops, as you pointed out, we still need to run the pre-condition checks and reject some of them.

Hope I didn't miss anything 😅

banach-space committed:
fixup! [mlir][linalg] Add a new helper hook - `isVectorizable`

Add tests to demonstrate the new functionality added to
`transform.structured.vectorize_children_and_apply_patterns`.

banach-space commented:
Note - I've added two tests for `tensor.pack` to demonstrate how `transform.structured.vectorize_children_and_apply_patterns` is being extended.

hanhanW commented:
I see your point, thanks for the details. It's very helpful! I'm generally +1 on removing the dup code between transform ops and pattern rewriting. It reduces the probability that one is supported in transform ops but not pattern rewriting.

I think `hasVectorizationImpl` is a better name. What do you think?

banach-space committed:
fixup! fixup! [mlir][linalg] Add a new helper hook - `isVectorizable`

Rename according to @hanhanW's suggestion

banach-space commented:
> I think hasVectorizationImpl is a better name. What do you think?

Agreed, thank you for the suggestion, updated :) 🙏🏻

hanhanW reviewed (approved):
LGTM, thanks! Please wait for an additional approval before landing it. Others could have different comments.

dcaballe commented:
LG. Just one comment: shouldn't we use that check within the vectorizer preconditions?

banach-space committed:
fixup! fixup! fixup! [mlir][linalg] Add a new helper hook - `isVectorizable`

Use hasVectorizationImpl in vectorizeOpPrecondition

banach-space commented:
@rengolin I will assume that you are OK with this and land tomorrow. Do let me know if my assumption is incorrect :)
